crates/rpc/rpc-builder/src/error.rs

```diff
@@ -0,0 +1,68 @@
+use std::net::SocketAddr;
+
+use jsonrpsee::core::Error as JsonRpseeError;
+use std::{io, io::ErrorKind};
+
+/// Rpc server kind.
+#[derive(Debug)]
+pub enum ServerKind {
+    /// Http.
+    Http(SocketAddr),
+    /// Websocket.
+    WS(SocketAddr),
+    /// Auth.
+    Auth(SocketAddr),
+    /// Unknown.
+    Unknown,
+}
+
+impl std::fmt::Display for ServerKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ServerKind::Http(addr) => write!(f, "{addr} (HTTP-RPC server)"),
+            ServerKind::WS(addr) => write!(f, "{addr} (WS-RPC server)"),
+            ServerKind::Auth(addr) => write!(f, "{addr} (AUTH server)"),
+            ServerKind::Unknown => write!(f, "(UNKNOWN server)"),
+        }
+    }
+}
+
+/// Rpc Errors.
+#[derive(Debug, thiserror::Error)]
+pub enum RpcError {
+    /// Wrapper for `jsonrpsee::core::Error`.
+    #[error(transparent)]
+    RpcError(JsonRpseeError),
+    /// Address already in use.
+    #[error("Address {kind} is already in use (os error 98)")]
+    AddressAlreadyInUse {
+        /// Server kind.
+        kind: ServerKind,
+        /// IO error.
+        error: io::Error,
+    },
+    /// Custom error.
+    #[error("{0}")]
+    Custom(String),
+}
+
+impl RpcError {
+    /// Converts a `jsonrpsee::core::Error` to a more descriptive `RpcError`.
+    pub fn from_jsonrpsee_error(err: JsonRpseeError, server_kind: Option<ServerKind>) -> RpcError {
+        match err {
+            JsonRpseeError::Transport(err) => match err.downcast::<io::Error>() {
+                Ok(io_error) => {
+                    if io_error.kind() == ErrorKind::AddrInUse {
+                        return RpcError::AddressAlreadyInUse {
+                            kind: server_kind.unwrap_or(ServerKind::Unknown),
```

Comment:
I see, but what I meant was put this entire downcast into an if let.

But looks like the `None` case is actually redundant. so we can:
* make this: `server_kind: ServerKind`

use `?` operator where currently `None` is used in the function call.

Reply:
oh yes, I totally missed the `#[from]` in the `RpcError(#[from] JsonRpseeError),` variant to get the `From` impl